---
title: "my title"
author: "my name (dummy@mail.com)"
date: "Last update: `r format(Sys.time(), '%d %B, %Y')`"      
output:
  html_document:
    toc: true
    fig_caption: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# Clustering

## **Optional exercise** 

Follow the tutorial from the [clustering section](https://genomicsclass.github.io/book/pages/clustering_and_heatmaps.html) of the book [PH525x series - Biomedical Data Science](http://genomicsclass.github.io/book/). 

Install the package [tissuesGeneExpression](https://github.com/genomicsclass/tissuesGeneExpression) by running in R: 

```{r, eval=F}
# install.packages("devtools")
# install.packages("rafalib")
# install_github("genomicsclass/tissuesGeneExpression")
# library(devtools)
# library(rafalib)
# library(tissuesGeneExpression)
# 
# 
# data(tissuesGeneExpression)
# d <- dist( t(e) )
# mypar()
# hc <- hclust(d)
# hc
# table(t(e))
# tissue
# plot(hc,labels=tissue,cex=0.5)
# myplclust(hc, labels=tissue, lab.col=as.fumeric(tissue), cex=0.5)
# abline(h=120)
# 
# hclusters <- cutree(hc, h=120)
# table(true=tissue, cluster=hclusters)
```

## Clustering gene expression data in healthy tissues

Download the [data](https://www.ebi.ac.uk/biostudies/arrayexpress/studies/E-MTAB-6081) (design and tpm files) corresponding to the publication [An RNASeq normal tissue atlas for mouse and rat](https://www.nature.com/articles/sdata2017185). 
Download the [gene expression data](https://storage.googleapis.com/adult-gtex/bulk-gex/v8/rna-seq/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_median_tpm.gct.gz)
corresponding to the publication  [The Genotype-Tissue Expression (GTEx) pilot analysis: multitissue gene regulation in humans](https://www.science.org/doi/10.1126/science.1262110) from  the [GTEX portal](https://gtexportal.org/home/datasets)

From GTEX data, keep only tissues belonging to the following categories:  


```{r}

# load data
human <- read.delim("data.gct", stringsAsFactors = FALSE, sep = "\t", skip = 2)
mouse <- read.delim("mouse_tpm.txt")
mouseD <- read.delim("mouse_design.txt")
rat <- read.delim("rat_tpm.txt")
ratD <- read.delim("rat_design.txt")

# convert cols to lower
colnames(human) <- tolower(colnames(human))
colnames(mouse) <- tolower(colnames(mouse))
colnames(rat) <- tolower(colnames(rat))


gtex_tissues <-  c("name", "colon", "ileum", "duodenum", "jejunum", "small intestine"  , "muscle", "pancreas", "liver", "stomach",  "kidney",  "quadriceps", "thymus", "heart" ,    "esophagus", "brain" )



# keep only the columns of interest

keep_subset <- function(tissues, columns){
  columns_to_keep_data <- c()
  
  for (t in tissues){
    for(a in columns){
      #if (any(grepl(t, columns_to_keep_data))) {
      #  next
      #}
      if (grepl(t, a)) {
        columns_to_keep_data <- c(columns_to_keep_data, a)
      }
    }
  }
  return (columns_to_keep_data)
}

human_cols <- keep_subset(gtex_tissues, colnames(human))


mouse_cols <- keep_subset(gtex_tissues, colnames(mouse))

rat_cols <- keep_subset(gtex_tissues, colnames(rat))



human <- human[human_cols]

# for some reason... if we add a dummy columns and later take them out, the columns with the same name will have
# number suffixes

mouse <- mouse[mouse_cols]
new_column <- rep(0, nrow(mouse))  
mouse <- cbind(dummy_column = new_column, mouse)


rat <- rat[rat_cols] 
new_column <- rep(0, nrow(rat))
rat <- cbind(dummy_column = new_column, rat)

# make the columns names unified... for example only brain... etc
unify_column_names <- function(df){
  new_column_names <- c()

  for (col in colnames(df)) {
    for (n in gtex_tissues){
      if (grepl(n, col)) {
        new_column_names <- c(new_column_names, n)
        break
      }
    }
  }
  return (new_column_names)
}

# put the unified column names in the dataframes

colnames(human) <- unify_column_names(human)
rownames(human) <- human$name
human <- human[,-1]


# here we have the use of dummy column... note the brain, brain.1, brain.2 etc.

colnames(mouse) <-c("dummy_column", unify_column_names(mouse))
mouse <- mouse[,-1]



colnames(rat) <- c("dummy_column", unify_column_names(rat))
rat <- rat[,-1]


# Keep only those columns in the datasets
cols_to_keep <- intersect(colnames(human), colnames(mouse))
cols_to_keep
human <- human[,cols_to_keep]
mouse <- mouse[,cols_to_keep]

rat <- rat[,cols_to_keep]


# Add suffixes to differentiate columns by the species
colnames(human) <- paste(colnames(human), "_human", sep = "")

colnames(mouse) <- paste(colnames(mouse), "_mouse", sep = "")

colnames(rat) <- paste(colnames(rat), "_rat", sep = "")

mapping <- read.csv("mapping.csv")
head(mapping)

# Add a new column ID
human$ID <- rownames(human)
rat$ID <- rownames(rat)
mouse$ID <- rownames(mouse)



# Create a one big dataframe with all the species
human_mapping <- merge(human, mapping, by.x = "ID", by.y = "Gene.stable.ID.version")

human_mapping_rat <- merge(human_mapping, rat, by.x = "Rat.gene.stable.ID", by.y = "ID")


human_mapping_rat_mouse <- merge(human_mapping_rat, mouse, by.x = "Mouse.gene.stable.ID", by.y = "ID")


```

**pro tip** Do not manually copy from the column names. Convert all column names from GTEX data to lower case, and split them appropriately.


Cluster tissues using gene expression data. Run k-means and hierarchical clustering. For each algorithm, determine the optimal number of clusters. 


```{r}
# install.packages("factoextra")
library(factoextra)
library(ggplot2)


# Hierarchical clustering
all_tables <- human_mapping_rat_mouse[-(1:2)]

all_tables_t <- t(all_tables)
all_tables_t <- as.data.frame(all_tables_t)

colnames(all_tables_t) <- all_tables_t[1,]
all_tables_t <- all_tables_t[-1,]


all_tables_t <- sapply(all_tables_t, as.numeric)


rownames(all_tables_t) <- colnames(all_tables[-1])

# Scale
all_tables_t_s <- scale(all_tables_t)

# Check for NaN or NA values in the dataframe
has_na <- any(is.na(all_tables_t_s))
has_na

# Remove NA
all_tables_t_s[is.na(all_tables_t_s)] <- 0
all_tables_matrix <- dist(all_tables_t_s, method = "euclidean")


# See the tree
hc_result <- hclust(all_tables_matrix, method = "complete" )
plot(hc_result, labels = rownames(all_tables_t), cex = 0.6)


# Color the tree into 3 groups
tree_coloured <- fviz_dend(hc_result, k = 3, k_colors = c("#E41A1C", "#377EB8", "#4DAF4A"),
          color_labels_by_k = TRUE, rect = TRUE, show_labels = TRUE, cex = 0.5) + 
      guides(color = "none")

tree_coloured


# KMeans

fviz_nbclust(all_tables_t_s, kmeans, method = "wss")

km <- kmeans(all_tables_t_s, centers = 3, iter.max=10, nstart = 25)
km$tot.withinss
km$cluster

table(km$cluster)
res_cluster <- fviz_cluster(object = km, data = all_tables_matrix)
res_cluster

```

Compare the clustering results using both methodologies, and with the tissues/species. Show the results of the final partitions as a table. 

```{r}

# Create a table to see in which cluster is which tissue
cluster_assignments <- cutree(hc_result, k = 3)

cluster_table <- data.frame(Dendrogram_cluster = cluster_assignments, Kmeans_cluster = km$cluster)
cluster_table <- cluster_table[order(cluster_table$Dendrogram_cluster), ]
cluster_table

```

Plot a heatmap of the 50 genes with top variance over all samples. Add the information about tissue groups and model (human, rat and mouse) as annotations in the heatmap*. 

```{r}

```



# Exercise 2: Dimensionality reduction 

## PCA 
With the gene expression for different tissues and models, perform a PCA on the data and visualize the results (PC1 and PC2, and also, PC3 ). Label the points in the plot with their respective tissues/models. 

```{r}

```


Visualize the data using the PC1 and PC2 again, but this time, color the observations by cluster, using the k means clusters, with k of your choice. Produce a caption for the plot


```{r}

```


What are the top 50 genes that contribute to the PC1? Are they the same genes that are more variable according to the exercise 1?


```{r}

```

## tSNE 

Perform t-SNE on the dataset and visualize the results. Test at least 2 perplexity values.



```{r}

```



# session info {.unnumbered}

```{r, results='asis',  echo=FALSE, message=FALSE }
sessionInfo()
```
